File: src/checks/workflows.ts
import { Check, CheckContext, CheckResult } from '../types.js';
import { createClient, DripError } from '../drip-client.js';

// Extended context type to store workflow ID across checks
type WorkflowContext = CheckContext & { workflowId?: string };

export const workflowCreateCheck: Check = {
  name: 'Workflow Create',
  description: 'Create workflow definition',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);

    try {
      const sdk = (client as unknown as {
        sdk: {
          createWorkflow?: (data: { name: string; description?: string }) => Promise<{ id: string; name: string }>;
        };
      }).sdk;

      if (!sdk.createWorkflow) {
        const duration = performance.now() - start;
        return {
          name: 'Workflow Create',
          success: true,
          duration,
          message: 'Skipped (createWorkflow not available)',
          details: 'The createWorkflow method is not available in the SDK',
        };
      }

      const result = await sdk.createWorkflow({
        name: `health-check-workflow-${Date.now()}`,
        slug: `health_check_workflow_${Date.now()}`,
        description: 'Test workflow created by health check',
      });
      const duration = performance.now() - start;

      // Store workflow ID for subsequent checks
      (ctx as WorkflowContext).workflowId = result.id;

      return {
        name: 'Workflow Create',
        success: true,
        duration,
        message: `Created: ${result.id}`,
        details: `Name: ${result.name}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Workflow Create',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Workflow Create',
        success: false,
        duration,
        message: err.message || 'Failed to create workflow',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const workflowListCheck: Check = {
  name: 'Workflow List',
  description: 'List all workflows',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);

    try {
      const sdk = (client as unknown as {
        sdk: {
          listWorkflows?: () => Promise<{ data: Array<{ id: string; name: string }> } | Array<{ id: string; name: string }>>;
        };
      }).sdk;

      if (!sdk.listWorkflows) {
        const duration = performance.now() - start;
        return {
          name: 'Workflow List',
          success: true,
          duration,
          message: 'Skipped (listWorkflows not available)',
          details: 'The listWorkflows method is not available in the SDK',
        };
      }

      const result = await sdk.listWorkflows();
      const duration = performance.now() - start;

      const workflows = Array.isArray(result) ? result : result.data || [];
      const count = workflows.length;

      // Store the first workflow ID if available
      if (count > 0 && !(ctx as WorkflowContext).workflowId) {
        (ctx as WorkflowContext).workflowId = workflows[0].id;
      }

      return {
        name: 'Workflow List',
        success: true,
        duration,
        message: `Found ${count} workflow(s)`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Workflow List',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Workflow List',
        success: false,
        duration,
        message: err.message || 'Failed to list workflows',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

File: python/checks/workflows.py
"""Workflow operation checks."""
import uuid
from ..types import Check, CheckContext, CheckResult
from ..drip_client import create_client


async def _workflow_create_check(ctx: CheckContext) -> CheckResult:
    """Create workflow definition."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'create_workflow'):
            return CheckResult(
                name="workflow_create",
                success=True,
                duration=0,
                message="Skipped (create_workflow not available)",
                details="The create_workflow method is not available in the SDK"
            )

        workflow_slug = f"health-check-{uuid.uuid4().hex[:8]}"
        result = client.create_workflow(
            name=f"Health Check Workflow {workflow_slug}",
            slug=workflow_slug,
            description="Test workflow created by health check"
        )

        workflow_id = getattr(result, 'id', str(result))
        ctx.workflow_id = workflow_id

        return CheckResult(
            name="workflow_create",
            success=True,
            duration=0,
            message=f"Created workflow {workflow_id}",
            details=f"Slug: {workflow_slug}"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="workflow_create",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="workflow_create",
            success=False,
            duration=0,
            message=f"Failed to create workflow: {e}"
        )


workflow_create_check = Check(
    name="workflow_create",
    description="Create workflow definition",
    run=_workflow_create_check
)


async def _workflow_list_check(ctx: CheckContext) -> CheckResult:
    """List all workflows."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'list_workflows'):
            return CheckResult(
                name="workflow_list",
                success=True,
                duration=0,
                message="Skipped (list_workflows not available)",
                details="The list_workflows method is not available in the SDK"
            )

        result = client.list_workflows()

        if hasattr(result, 'data'):
            count = len(result.data)
            if count > 0 and not ctx.workflow_id:
                ctx.workflow_id = result.data[0].id
        elif isinstance(result, list):
            count = len(result)
            if count > 0 and not ctx.workflow_id:
                ctx.workflow_id = result[0].id if hasattr(result[0], 'id') else result[0].get('id')
        else:
            count = 1

        return CheckResult(
            name="workflow_list",
            success=True,
            duration=0,
            message=f"Found {count} workflow(s)"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="workflow_list",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="workflow_list",
            success=False,
            duration=0,
            message=f"Failed to list workflows: {e}"
        )


workflow_list_check = Check(
    name="workflow_list",
    description="List all workflows",
    run=_workflow_list_check
)
